python/ray/air/execution/_internal/event_manager.py

```diff
@@ -163,40 +234,452 @@ def wait(
+        for resource_request in self._resource_request_to_tracked_actors:
+            if max_actors and started_actors >= max_actors:
+                break
+
+            # While we have resources ready and there are actors left to schedule
+            while (
+                self._resource_manager.has_resources_ready(resource_request)
+                and self._resource_request_to_tracked_actors[resource_request]
+            ):
+                # Acquire resources for actor
+                acquired_resources = self._resource_manager.acquire_resources(
+                    resource_request
+                )
+                assert acquired_resources
+
+                # Get tracked actor to start
+                candidate_actors = self._resource_request_to_tracked_actors[
+                    resource_request
+                ]
+                assert candidate_actors
+
+                tracked_actor = candidate_actors.pop(0)
+
+                # Get actor class and arguments
+                actor_cls, kwargs, _ = self._tracked_actors_to_attrs.pop(tracked_actor)
+
+                if not isinstance(actor_cls, ray.actor.ActorClass):
+                    actor_cls = ray.remote(actor_cls)
+
+                # Associate to acquired resources
+                [remote_actor_cls] = acquired_resources.annotate_remote_entities(
+                    [actor_cls]
+                )
+
+                # Start Ray actor
+                actor = remote_actor_cls.remote(**kwargs)
+
+                self._tracked_actors_to_ray_actors_resources[tracked_actor] = (
+                    actor,
+                    acquired_resources,
+                )
+
+                # Enqueue cached futures
+                cached_tasks = self._tracked_actors_to_pending_actor_tasks.pop(
+                    tracked_actor, []
+                )
+                for tracked_actor_task, method_name, args, kwargs in cached_tasks:
+                    self._schedule_tracked_actor_task(
+                        tracked_actor_task=tracked_actor_task,
+                        method_name=method_name,
+                        args=args,
+                        kwargs=kwargs,
+                    )
+
+                # Trigger on_start callback
+                if tracked_actor._on_start:
+                    tracked_actor._on_start(tracked_actor)
+
+                started_actors += 1
+
+        return started_actors
+
+    def _try_stop_actors(self, max_actors: Optional[int] = None) -> int:
+        """Try to stop up to ``max_actors`` actors."""
+        stopped_actors = 0
+
+        for tracked_actor in list(self._tracked_actors_to_terminate):
+            if max_actors and stopped_actors >= max_actors:
+                break
+
+            if self._try_stop_tracked_actor(tracked_actor):
+                stopped_actors += 1
+
+        return stopped_actors
+
+    def _try_stop_tracked_actor(self, tracked_actor: TrackedActor) -> bool:
```

Comment:
The important thing to keep in mind here is that we always cache actor termination. This function call be called multiple times for the same tracked actor in different calls to wait().

Reply:
maybe Ray core can tell you an ActorHandle is invalid, so we don't have to cache these things?